### PR TITLE
pydantic v2

### PR DESCRIPTION
**Pull request by awoimbee**
I may not have the time to complete this PR, feel free to create a new one using this one as inspiration/base !

### CODE REVIEW

**Comment by mansenfranzen** [comment]
Hi @awoimbee,

thanks for the PR draft! 

I have not run any tests yet but the changes look rather minor and yet promising so far. I hope to get some time off to make progress with the pydantic v2 support. I'm happy for any more contribution :-).

Greetings

**Comment by Galarzaa90** [comment]
I tried to check this out, I ran all unit tests, they were okay.

But I couldn't build the docs to test that. Had a hard time with graphviz (on Windows)

File: tests/roots/test-base/target/edgecase_model_as_attr.py [hunk 1]
```diff
@@ -14,7 +14,7 @@ class Model(BaseModel):
     field2: str
     """Field2 Doc String"""
 
-    @validator("field")
+    @field_validator("field")
```
Comment (PriOliveira) [inline tests/roots/test-base/target/edgecase_model_as_attr.py:17]: I was taking a quick look and noticed that the import is missing being updated as well, same in some other files in the tests (and some docs)

**Comment by mansenfranzen** [comment]
@awoimbee Thanks so much for continuing your work on this PR!

I will have some time the next days to support you. Could you allow me to push on your branch? I just tried but I don't seem to have permissions.

**Comment by mansenfranzen** [comment]
It works, I can push now. 

Regarding `PydanticConfigClassDocumenter`: This was an exotic feature and as far as I can see, it is not even needed by the model/settings documenter. Hence, since there are many changes for v2, I suggest to remove it to reduce complexity and maintenance costs.

File: pyproject.toml [hunk 2]
```diff
@@ -18,9 +18,10 @@ classifiers = [
 include = ["sphinxcontrib/autodoc_pydantic/css/autodoc_pydantic.css"]
 
 [tool.poetry.dependencies]
-python = ">=3.7.1,<4.0.0"
-Sphinx = ">=3.4"
-pydantic = ">=1.5,<2.0.0"
+python = ">=3.8.0,<4.0.0"
```
Comment (mansenfranzen) [inline pyproject.toml:21]: Was there a specific reason to remove support for python 3.7? It is EOL by June 2023, however I would like to keep backwards compatibility if possible.

Comment (awoimbee) [reply pyproject.toml:21]: I thought pydantic would remove py37 support after seeing https://github.com/pydantic/pydantic/issues/6426 but they fixed the issue. I'll add py37 back, there is a workaround needed around `importlib_metadata` -> `importlib.metadata`

File: sphinxcontrib/autodoc_pydantic/inspection.py [hunk 3]
```diff
@@ -532,21 +531,21 @@ def get_field_validator_mapping(self) -> Dict[str, List[ValidatorAdapter]]:
...
-        for func in self.model.__pre_root_validators__:
-            mapping["*"].append(ValidatorAdapter(func=func,
-                                                 root_pre=True))
-
-        # root post
-        for _, func in self.model.__post_root_validators__:
-            mapping["*"].append(ValidatorAdapter(func=func,
-                                                 root_post=True))
+        # root validators
+        for validator in model_decorators.root_validators.values():
```
Comment (PriOliveira) [inline sphinxcontrib/autodoc_pydantic/inspection.py:545]:
```suggestion
        for validator in model_decorators.model_validators.values():
```

File: tests/roots/test-base/target/configuration.py [hunk 4]
```diff
@@ -1,6 +1,7 @@
 from typing import Optional
 
-from pydantic import BaseModel, validator, Field, BaseSettings, root_validator
+from pydantic import BaseModel, field_validator, Field, model_validator, ConfigDict, root_validator
```
Comment (PriOliveira) [inline tests/roots/test-base/target/configuration.py:3]:
```suggestion
from pydantic import BaseModel, field_validator, Field, model_validator, ConfigDict
```

File: tests/roots/test-base/target/configuration.py [hunk 5]
```diff
@@ -345,17 +332,17 @@ class ValidatorAsteriskRootValidator(BaseModel):
...
     field: int = 1
 
-    @validator("*")
+    @field_validator("*")
     def check(cls, v):
         """Check."""
         return v
 
-    @root_validator
+    @root_validator(skip_on_failure=True)
```
Comment (PriOliveira) [inline tests/roots/test-base/target/configuration.py:340]:
```suggestion
    @model_validator
```

File: tests/roots/test-base/target/example_generics.py [hunk 6]
```diff
@@ -24,7 +24,7 @@ class Response(GenericModel, Generic[DataT]):
     data: Optional[DataT]
     error: Optional[Error]
 
-    @validator('error', always=True)
+    @field_validator('error', always=True)
```
Comment (PriOliveira) [inline tests/roots/test-base/target/example_generics.py:27]: By the migration docs, it seems `field_validator` doesn't have `always` anymore, but I couldn't find what would be equivalent

File: tests/roots/test-base/target/example_validators.py [hunk 7]
```diff
@@ -1,4 +1,4 @@
-from pydantic import BaseModel, validator, root_validator
+from pydantic import BaseModel, field_validator, root_validator
```
Comment (PriOliveira) [inline tests/roots/test-base/target/example_validators.py:1]:
```suggestion
from pydantic import BaseModel, field_validator, model_validator
```

File: tests/roots/test-base/target/usage_autosummary.py [hunk 8]
```diff
@@ -29,10 +30,7 @@ def check_max_length_ten(cls, v):
         if not len(v) < 10:
             raise ValueError("No more than 10 characters allowed")
 
-    class Config:
-        env_prefix = "foo_"
-        allow_mutation = True
-
+    model_config = ConfigDict(frozen=False)
```
Comment (PriOliveira) [inline tests/roots/test-base/target/usage_autosummary.py:33]:
```suggestion
    model_config = SettingsConfigDict(env_prefix="foo_", frozen=False)
```

File: tests/roots/test-events-add-css-fallback/add_css_fallback.py [hunk 9]
```diff
@@ -29,10 +30,7 @@ def check_max_length_ten(cls, v):
         if not len(v) < 10:
             raise ValueError("No more than 10 characters allowed")
 
-    class Config:
-        """Dummy DocString"""
-        env_prefix = "foo_"
-        allow_mutation = True
+    model_config = ConfigDict(frozen=False)
```
Comment (PriOliveira) [inline tests/roots/test-events-add-css-fallback/add_css_fallback.py:33]:
```suggestion
    model_config = SettingsConfigDict(env_prefix="foo_", frozen=False)
```

**Review by PriOliveira — COMMENTED** [review]
I was taking a look at the changes, even though I know it is still a draft, hopefully you don't mind for the comments about some stuff I found along the way


Also in the end the README will also need to be updated, for example:
> Supports pydantic >= 1.5.0, < 2.0.0 and sphinx >= 3.4.0

**Comment by mansenfranzen** [comment]
> I was taking a look at the changes, even though I know it is still a draft, hopefully you don't mind for the comments about some stuff I found along the way
> 
> Also in the end the README will also need to be updated, for example:
> 
> > Supports pydantic >= 1.5.0, < 2.0.0 and sphinx >= 3.4.0

Thanks @PriOliveira - all comments are more than appreciated.

**Comment by mansenfranzen** [comment]
I'm going merge this PR into the new "main-2.x" branch even though it is still WIP. Almost all tests have passed now. I will generate the new docs based on pydantic v2 which help to further improve the docs and update outdated information.

**Pull request merged**